src/main/java/gregicadditions/machines/TileEntityProcessingArray.java

```diff
@@ -417,35 +471,195 @@ protected boolean setupAndConsumeRecipeInputs(Recipe recipe) {
 				recipe.matches(true, importInventory, importFluids);
 		}
 
+		protected boolean didMachinesChange(ItemStack newMachineStack) {
+
+			if(newMachineStack == null || this.oldMachineStack == null) {
+				return true;
+			}
+
+			return !ItemStack.areItemStacksEqual(this.oldMachineStack, newMachineStack);
+		}
+
 		@Override
 		protected void trySearchNewRecipe() {
+			if(metaTileEntity instanceof TileEntityProcessingArray && ((TileEntityProcessingArray) metaTileEntity).isDistinct) {
+				trySearchNewRecipeDistinct();
+			}
+			else {
+				trySearchNewRecipeCombined();
+			}
+		}
+
+		private void trySearchNewRecipeCombined() {
 			long maxVoltage = getMaxVoltage();
-			Recipe currentRecipe = null;
+			Recipe currentRecipe;
 			IItemHandlerModifiable importInventory = getInputInventory();
 			IMultipleTankHandler importFluids = getInputTank();
 
+			ItemStack newMachineStack = findMachineStack();
+
 			boolean dirty = checkRecipeInputsDirty(importInventory, importFluids);
 			if(dirty || forceRecipeRecheck) {
-				this.forceRecipeRecheck = false;
+				//Check if the machine that the PA is operating on has changed
+				if(didMachinesChange(newMachineStack)) {
+					previousRecipe = null;
+					oldMachineStack = null;
+				}
+			}
 
-				// else, try searching new recipe for given inputs
+			if(previousRecipe != null &&
+					previousRecipe.matches(false, importInventory, importFluids)) {
+				currentRecipe = previousRecipe;
+			}
+			else {
+				//If the previous recipe was null, or does not match the current recipe, search for a new recipe
 				currentRecipe = findRecipe(maxVoltage, importInventory, importFluids);
```

Comment:
I have changed things so that `findMachineStack` is not called twice, although I am a little confused as why `newMachineStack = machineItemStack` would not be needed any more. Would you mind explaining the logic behind that?

Reply:
i assumed that the reason `newMachineStack = machineItemStack` was there in the first place is that the `findMachineStack` call in `findRecipe` could somehow find a different stack that then had to be written back to `newMachineStack`, but since `findMachineStack` is only called once and its value assigned to both variables, there shouldn't be a way they can diverge any more

Reply:
(i'm not even sure they could ever diverge in the original code...)